api/src/main/java/org/openmrs/module/legacyui/api/impl/LegacyUIImpl.java

```diff
@@ -0,0 +1,291 @@
+		
+		// Check input parameters
+		if (patient == null) {
+			throw new APIException("Attempting to convert state of an invalid patient");
+		}
+		if (trigger == null) {
+			throw new APIException("Attempting to convert state for a patient without a valid trigger concept");
+		}
+		if (dateConverted == null) {
+			throw new APIException("Invalid date for converting patient state");
+		}
+		
+		for (PatientProgram patientProgram : Context.getProgramWorkflowService().getPatientPrograms(patient, null, null,
+		    null, null, null, false)) {
+			//skip past patient programs that already completed
+			if (patientProgram.getDateCompleted() == null) {
+				Set<ProgramWorkflow> workflows = patientProgram.getProgram().getWorkflows();
+				for (ProgramWorkflow workflow : workflows) {
+					// (getWorkflows() is only returning over nonretired workflows)
+					PatientState patientState = patientProgram.getCurrentState(workflow);
+					
+					// #1080 cannot exit patient from care  
+					// Should allow a transition from a null state to a terminal state
+					// Or we should require a user to ALWAYS add an initial workflow/state when a patient is added to a program
+					ProgramWorkflowState currentState = (patientState != null) ? patientState.getState() : null;
+					ProgramWorkflowState transitionState = workflow.getState(trigger);
+					
+					log.debug("Transitioning from current state [" + currentState + "]");
+					log.debug("|---> Transitioning to final state [" + transitionState + "]");
+					
+					if (transitionState != null && workflow.isLegalTransition(currentState, transitionState)) {
+						patientProgram.transitionToState(transitionState, dateConverted);
+						log.debug("State Conversion Triggered: patientProgram=" + patientProgram + " transition from "
+						        + currentState + " to " + transitionState + " on " + dateConverted);
+					}
+				}
+				
+				// #1068 - Exiting a patient from care causes "not-null property references
+				// a null or transient value: org.openmrs.PatientState.dateCreated". Explicitly
+				// calling the savePatientProgram() method will populate the metadata properties.
+				// 
+				// #1067 - We should explicitly save the patient program rather than let 
+				// Hibernate do so when it flushes the session.
+				Context.getProgramWorkflowService().savePatientProgram(patientProgram);
+			}
+		}
+	}
+	
+	/**
+	 * @see OrderExtensionService#getProviderForUser(User)
+	 */
+	public Provider getProviderForUser(User user) {
+		ProviderService ps = Context.getProviderService();
+		Collection<Provider> providers = ps.getProvidersByPerson(user.getPerson(), true);
+		if (providers.isEmpty()) {
+			throw new IllegalStateException("User " + user + " has no provider accounts, unable to create orders");
+		}
+		return providers.iterator().next();
+	}
+	
+	/**
+	 * Copied from OpenMRS core 1.9.x
+	 * See https://github.com/openmrs/openmrs-core/blob/1.9.x/api/src/main/java/org/openmrs/order/OrderUtil.java#L52
```

Comment:
Same as above for using tags

Reply:
[Fixed](https://github.com/openmrs/openmrs-module-legacyui/pull/128/commits/9dd92c8a5560377daedb2e231c639e315fa177b8)